src/chemistry_functionality.jl

```diff
@@ -0,0 +1,211 @@
+function component_coefficients(s)
+    return [c => co for (c, co) in zip(components(s), coefficients(s))]
+end
+
+
+### Balancing Code
+
+function create_matrix(reaction::Catalyst.Reaction)
+    compounds = [reaction.substrates; reaction.products]
+    atoms = [] # Array to store unique atoms
+    n_atoms = 0
+    A = zeros(Int, 0, length(compounds))
+
+    for (j, compound) in enumerate(compounds)
+        # Check if the compound is a valid compound
+        if iscompound(compound)
+            # Get component coefficients of the compound
+            pairs = component_coefficients(compound)
+            if pairs == Nothing 
+                continue
+            end
+        else 
+            # If not a compound, assume coefficient of 1
+            pairs = [(compound, 1)]
+        end
+
+        for pair in pairs
+            # Extract atom and coefficient from the pair
+            atom, coeff = pair
+            i = findfirst(x -> isequal(x, atom), atoms)
+            if i === nothing  
+                # Add the atom to the atoms array if it's not already present
+                push!(atoms, atom)
+                n_atoms += 1
+                A = [A; zeros(Int, 1, length(compounds))]
+                i = n_atoms
+            end
+            # Adjust coefficient based on whether the compound is a product or substrate
+            coeff = any(map(p -> isequal(p, compounds[j]), reaction.products)) ? -coeff : coeff
+            A[i, j] = coeff
+        end
+    end
+
+    return A
+end
+
+function get_stoich(reaction::Reaction)
+    # Create the matrix A using create_matrix function.
+    A = create_matrix(reaction)
+    
+    m, n = size(A)
+        if m == n
+            B = zeros(Int64, size(A,1))
+        else
+            # Append a row with last element as 1 
+            new_row = zeros(Int, 1, size(A, 2))
+            new_row[end] = 1
+            A = vcat(A, new_row)
+
+            B = zeros(Int64, size(A,1))
+            B[end] = 1
+        end
+
+    # Concatenate B to A to form an augmented matrix   
+    AB = [A B]
+
+    # Apply the Bareiss algorithm
+    ModelingToolkit.bareiss!(AB)
+
+    # Extract the transformed A and B
+    A_transformed = AB[:, 1:end-1]
+    B_transformed = AB[:, end]
+
+    # Convert A_transformed to rational numbers
+    A_transformed_rational = Rational.(A_transformed)
+
+    # Convert B_transformed to rational numbers
+    B_transformed_rational = Rational.(B_transformed)
+
+    # Perform backward substitution
+    X = backward_substitution(A_transformed_rational,B_transformed_rational)
+
+    # Get the denominators of the rational numbers in X
+    denominators = denominator.(X)
+
+    # Compute the LCM of the denominators
+    lcm_value = reduce(lcm, denominators)
+
+    # Multiply each element in X by the LCM of the denominators
+    X_multiplied = X .* lcm_value
+
+    # Convert the rational numbers to integers
+    X_integers = numerator.(X_multiplied)
+
+    return abs.(X_integers)
```

Comment:
@LalitChauhan56 do you have a reference for this approach? When do you only add one row when m != n? What if m = n + k with k > 1?

More generally, why not use the `nullspace` function from ModelingToolkit that we use for conservation laws, see https://github.com/SciML/Catalyst.jl/blob/e91b001dbaa9ccfe855a54ed79661b332d6b05ae/src/networkapi.jl#L1200-L1241 to get an integer nullspace basis?